File: ios/Anemone SDK/general/client/LocalHttpClient.swift
//
//  LocalHttpClient.swift
//  Anemone SDK
//
//  Created by Manuel Gonzalez Villegas on 5/12/16.
//  Copyright © 2016 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public class LocalHttpClient: IHttpClient {
    public init() { }

    public func request(_ httpRequest: HttpRequest) -> Promise<HttpResponse> {
        Promise(executeInBackground: true) { fulfill, reject in
            let resource = "assets/\(httpRequest.url)"

            if
                let path = Bundle(for: type(of: self)).path(forResource: resource, ofType: "json"),
                let data = try? Data(contentsOf: URL(fileURLWithPath: path))
            {
                fulfill(HttpResponse(request: httpRequest, headers: [:], code: 200, body: data))
            } else {
                reject(
                    AnemoneException.network(
                        response: HttpResponse(request: httpRequest, headers: [:], code: 404, body: Data()),
                        error: AnemoneException.notFound,
                        serverMessage: nil
                    )
                )
            }
        }
    }
}

File: android/app/src/main/java/com/playtomic/test/anemone_sdk/general/client/LocalHttpClient.kt
package com.anemonesdk.general.client

import android.content.Context
import com.anemonesdk.general.exception.AnemoneException
import com.playtomic.foundation.promise.Promise

/**
 * Created by mgonzalez on 21/12/16.
 */

class LocalHttpClient(private val context: Context) : IHttpClient {

    override fun request(httpRequest: HttpRequest): Promise<HttpResponse> {
        var endpoint = httpRequest.url
        if (endpoint.startsWith("/")) {
            endpoint = endpoint.substring(1)
        }
        val path = endpoint + ".json"

        return Promise(executeInBackground = true) { fulfill, reject ->
            try {
                val inputStream = context.assets.open(path)
                val data = inputStream.readBytes()
                inputStream.close()
                fulfill(HttpResponse(request = httpRequest, headers = mapOf(), code = 200, body = data))
            } catch (e: Exception) {
                reject(
                    AnemoneException.network(
                        response = HttpResponse(request = httpRequest, headers = mapOf(), code = 404, body = ByteArray(0)),
                        error = e,
                        serverMessage = null
                    )
                )
            }
        }
    }
}
